src/python/pants/util/collections.py

```diff
@@ -71,3 +74,49 @@ def ensure_str_list(val: str | Iterable[str], *, allow_single_str: bool = False)
     If `allow_single_str` is True, a single `str` will be wrapped into a `List[str]`.
     """
     return ensure_list(val, expected_type=str, allow_single_scalar=allow_single_str)
+
+
+def partition_sequentially(
+    items: Sequence[_T],
+    *,
+    key: Callable[[_T], str],
+    size_min: int,
+    size_max: int | None = None,
+) -> Iterator[list[_T]]:
+    """Stably partitions the given items into batches of at least size_min.
+
+    The "stability" property refers to avoiding adjusting all batches when a single item is added,
+    which could happen if the items were trivially windowed using `itertools.islice` and an
+    item was added near the front of the list.
+
+    Batches will be capped to `size_max`, which defaults `size_min*2`.
+    """
+
+    # To stably partition the arguments into ranges of at least `size_min`, we sort them, and
+    # create a new batch sequentially once we have the minimum number of entries, _and_ we encounter
+    # an item hash prefixed with a threshold of zeros.
```

Comment:
I'm definitely missing what the leading zeros check is about. You explain trying to accommodate adding items disturbing batches minimally, but I don't understand the mechanism of how this helps. Is it tied to characteristics of the Rust hash function used? Maybe a unit test of this function that shows how adding an item to an N bucket chain only results in ~1 bucket changing contents?

Reply:
+1 to unit tests of this function.

Reply:
I can sort of see what's going on here, but some comments explaining it would be really helpful. Especially justifying the selection of zero_prefix_threshold .

Reply:
Expanded the comment and added a test.